### PR TITLE
Jd/compute improvements

### PR DESCRIPTION
**Pull request by jd-lara**
Thanks for opening a PR to PowerNetworkMatrices.jl, please take note of the following when making a PR:

Check the [contributor guidelines](https://nrel-siip.github.io/PowerNetworkMatrices.jl/stable/code_base_developer_guide/developer/)

cc @alefcastelli

### CODE REVIEW

File: src/BA_ABA_matrices.jl [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# ABA matrix #################################################################
+struct ABA_Matrix <: PowerNetworkMatrix{Float64}
+    data::SparseArrays.SparseMatrixCSC{Float64, Int32}
+end
+
+# create the ABA matrix
+function ABA_Matrix(A::SparseArrays.SparseMatrixCSC{Int8, Int32},
+    BA::SparseArrays.SparseMatrixCSC{T, Int32}
+    where {T <: Union{Float32, Float64}})
+    data = calculate_ABA_matrix(A, BA)
```
Comment (jd-lara) [inline src/BA_ABA_matrices.jl:22]: The call to this method is incorrect. It requires passing the slacks and those aren't passed here.

File: src/BA_ABA_matrices.jl [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    return BA_Matrix(data)
+end
+
+# ABA matrix #################################################################
+struct ABA_Matrix <: PowerNetworkMatrix{Float64}
+    data::SparseArrays.SparseMatrixCSC{Float64, Int32}
+end
+
+# create the ABA matrix
+function ABA_Matrix(A::SparseArrays.SparseMatrixCSC{Int8, Int32},
```
Comment (jd-lara) [inline src/BA_ABA_matrices.jl:19]: This function isn't tested. Before merging this code needs tests

File: test/temp_main.jl [hunk 3]
```diff
@@ -0,0 +1,31 @@
+using PowerSystemCaseBuilder
+using PowerSystems
+
+const PNM = include("/Users/acastel2/Documents/GitHub/PowerNetworkMatrices.jl/src/PowerNetworkMatrices.jl")
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply test/temp_main.jl]: This file needs to be removed. There is a reference to you local files.

File: src/ptdf_calculations.jl [hunk 4]
```diff
@@ -22,74 +22,35 @@ function binfo_check(binfo::Int)
...
+    elseif linear_solver == "Dense"
+        ptdf, a = calculate_PTDF_matrix_DENSE(branches, nodes, dist_slack)
+    elseif linear_solver == "MKLPardiso"
+        error("MKLPardiso still to be implemented")
     end
 
-    slacks =
-        [bus_lookup[PSY.get_number(n)] for n in nodes if PSY.get_bustype(n) == BusTypes.REF]
-    isempty(slacks) && error("System must have a reference bus!")
+    return ptdf, a
```
Comment (jd-lara) [inline src/ptdf_calculations.jl:37]: there is no need to return a here anymore

File: src/common.jl [hunk 5]
```diff
@@ -1,269 +1,250 @@
...
+
+    return BA
+end
+
+# ABA matrix evaluation ######################################################
+function calculate_ABA_matrix(
+    A::SparseArrays.SparseMatrixCSC{Int8, Int32},
+    BA::SparseArrays.SparseMatrixCSC{T, Int32} where {T <: Union{Float32, Float64}},
+    slack_positions::Vector{Int64})
+    return A[:, setdiff(1:end, slack_positions[1])]' * BA
```
Comment (jd-lara) [inline src/common.jl:102]: user `matmul!` here.

File: src/common.jl [hunk 6]
```diff
@@ -1,269 +1,250 @@
...
+    ldiv!(ABA_inv, K, Ix)
+
+    slack_position = slack_positions[1]
+    if dist_slack[1] == 0.1 && length(dist_slack) == 1
+        S = BA * ABA_inv
+        @views S =
+            hcat(S[:, 1:(slack_position - 1)], zeros(linecount), S[:, slack_position:end])
+    elseif dist_slack[1] != 0.1 && length(dist_slack) == buscount
+        @info "Distributed bus"
+        S = BA * ABA_inv
```
Comment (jd-lara) [inline src/common.jl:127]: use `matmul!` here

File: src/common.jl [hunk 7]
```diff
@@ -1,269 +1,250 @@
...
+    BA = calculate_BA_matrix(branches, slack_positions, bus_lookup)
+    ABA = calculate_ABA_matrix(A, BA, slack_positions)
+    K = klu(ABA)
+    Ix = Matrix(1.0LinearAlgebra.I, size(ABA, 1), size(ABA, 1))
+    ABA_inv = zeros(Float64, size(ABA, 1), size(ABA, 2))
+    ldiv!(ABA_inv, K, Ix)
+
+    slack_position = slack_positions[1]
+    if dist_slack[1] == 0.1 && length(dist_slack) == 1
+        S = BA * ABA_inv
```
Comment (jd-lara) [inline src/common.jl:122]: same here.

File: src/common.jl [hunk 8]
```diff
@@ -1,269 +1,250 @@
...
+        @views S =
+            hcat(S[:, 1:(slack_position - 1)], zeros(linecount), S[:, slack_position:end])
```
Comment (jd-lara) [inline src/common.jl:124]: the hcat operation here is allocating a new matrix

File: src/common.jl [hunk 9]
```diff
@@ -1,269 +1,250 @@
...
     end
-    =#
-    return A.data[to_index(A, row, column)...]
-end
-Base.getindex(A::PowerNetworkMatrix, idx::CartesianIndex) = A.data[idx]
-Base.setindex!(A::PowerNetworkMatrix, v, idx...) = A.data[to_index(A, idx...)...] = v
-Base.setindex!(A::PowerNetworkMatrix, v, idx::CartesianIndex) = A.data[idx] = v
 
-Base.IndexStyle(::Type{PowerNetworkMatrix}) = IndexAnyCartesian()
+    return S, A
```
Comment (jd-lara) [inline src/common.jl:140]: no need to return A here

File: src/common.jl [hunk 10]
```diff
@@ -1,269 +1,250 @@
...
+    BA::SparseArrays.SparseMatrixCSC{T, Int32}
+    where {T <: Union{Float32, Float64}})
+    ABA = calculate_BA_matrix(A, BA)
+    Ix = Matrix(1.0I, size(ABA, 1), size(ABA, 1))
+    S = zeros(Float64, size(ABA, 1), size(ABA, 2))
+    ldiv!(S, K, Ix)
 
-struct PowerNetworkMatrixKey{T <: Tuple}
-    I::T
+    return BA * S
```
Comment (jd-lara) [inline src/common.jl:151]: use `matmul!` we know the size of the matrix here

File: src/common.jl [hunk 11]
```diff
@@ -1,269 +1,250 @@
...
-Base.setindex!(A::PowerNetworkMatrix, v, idx::CartesianIndex) = A.data[idx] = v
 
-Base.IndexStyle(::Type{PowerNetworkMatrix}) = IndexAnyCartesian()
+    return S, A
+end
 
-########
-# Keys #
-########
+function calculate_PTDF_matrix_KLU(A::SparseArrays.SparseMatrixCSC{Int8, Int32},
```
Comment (jd-lara) [inline src/common.jl:143]: This function isn't tested

File: src/ptdf_calculations.jl [hunk 12]
```diff
@@ -22,74 +22,35 @@ function binfo_check(binfo::Int)
...
-        @warn("Slack bus not identified in the Bus/Nodes list, can't build PTDF")
-        S = Array{Float64, 2}(undef, linecount, buscount)
-    else
-        @assert false
+function _buildptdf(A::SparseArrays.SparseMatrixCSC{Int8, Int32},
+    BA::SparseArrays.SparseMatrixCSC{T, Int32}
+    where {T <: Union{Float32, Float64}},
+    dist_slack::Vector{Float64},
+    linear_solver::String)
+    if linear_solver == "KLU"
```
Comment (jd-lara) [inline src/ptdf_calculations.jl:45]: we don't need A here.

File: test/temp_main.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (alefcastelli) [reply test/temp_main.jl]: file has been removed in a previous commit

File: src/BA_ABA_matrices.jl (hunk 1 above)
Comment (alefcastelli) [reply src/BA_ABA_matrices.jl:22]: You are right. Solved

File: src/ptdf_calculations.jl (hunk 4 above)
Comment (alefcastelli) [reply src/ptdf_calculations.jl:37]: we agreed to keep both for the moment

File: src/common.jl (hunk 9 above)
Comment (alefcastelli) [reply src/common.jl:140]: we agreed to keep them both for the momen

File: src/ptdf_calculations.jl (hunk 12 above)
Comment (alefcastelli) [reply src/ptdf_calculations.jl:45]: we agreed to keep them both for the momen

File: src/common.jl (hunk 6 above)
Comment (alefcastelli) [reply src/common.jl:127]: solved

File: src/common.jl (hunk 7 above)
Comment (alefcastelli) [reply src/common.jl:122]: solved

File: src/common.jl (hunk 10 above)
Comment (alefcastelli) [reply src/common.jl:151]: solved

File: src/common.jl (hunk 8 above)
Comment (alefcastelli) [reply src/common.jl:124]: solved

File: test/test_network_matrices.jl [hunk 13]
```diff
@@ -435,24 +435,27 @@ Ybus3_matpower[3, 3] = 1.379310344827586 - 16.351724137931036im
...
-    end
-
-    PRTS = PTDF(RTS)
-    bnums = sort([PSY.get_number(b) for b in PSY.get_components(Bus, RTS)])
-    for (ibr, br) in enumerate(RTS_branchnames), (ib, b) in enumerate(bnums)
-        @test getindex(PRTS, br, b) - SRTS_GMLC[ibr, ib] <= 1e-3
+    for solver in ["KLU", "Dense"]
+        P5 = PTDF(branches_5, nodes_5; linear_solver = solver)
+        @test maximum(P5.data - S5_slackB4) <= 1e-3
+        @test abs(P5[branches_5[1], nodes_5[1]] - 0.1939166051164976) <= 1e-9
```
Comment (jd-lara) [inline test/test_network_matrices.jl:441]: should use here the function `isapprox()`

File: test/test_network_matrices.jl [hunk 14]
```diff
@@ -435,24 +435,27 @@ Ybus3_matpower[3, 3] = 1.379310344827586 - 16.351724137931036im
...
+        # @test P5[branches_5[1], nodes_5[1]] == 0.1939166051164976
+
+        nodes_14 = nodes14()
+        branches_14 = branches14(nodes_14)
+        P14 = PTDF(branches_14, nodes_14)
+        @test maximum(P14.data - S14_slackB1) <= 1e-3
+
+        P5NS = PTDF([branches_5[b] for b in Br5NS_ids], [nodes_5[b] for b in Bu5NS_ids])
+        for br in Br5NS_ids, b in Bu5NS_ids
+            @test getindex(P5NS, string(br), b) - S5_slackB4[br, b] <= 1e-3
```
Comment (jd-lara) [inline test/test_network_matrices.jl:451]: same here. This should have been using `isapprox()`

File: test/test_network_matrices.jl [hunk 15]
```diff
@@ -461,7 +464,8 @@ end
     branches_5 = branches5(nodes_5)
     L5 = LODF(branches_5, nodes_5)
     @test maximum(L5.data - Lodf_5) <= 1e-3
-    @test L5[branches_5[1], branches_5[2]] == 0.3447946513849091
+    @test abs(L5[branches_5[1], branches_5[2]] - 0.3447946513849091) < 1e-9
```
Comment (jd-lara) [inline test/test_network_matrices.jl:467]: Same here

**Pull request merged**